client/src/api/v3/boost.ts

```diff
@@ -1,28 +1,121 @@
- * @param configFile file containing the configuration content
+ * @param name name of the config file
+ * @param configContent configuration content
+ */
+function sendConfig(
+  actionType: payloadAction,
+  type: ConfigT,
+  name: string,
+  configContent: string | null,
+) {
+  const channel = 'send-config';
+  const payload = {
+    action: actionType,
+    configType: type,
+    fileName: name,
+    content: configContent,
+  };
+  // To check things work without needing the server started
+  console.log(payload);
+  emit(channel, JSON.stringify(payload));
+}
+
+const configTypeToFileSuffix: { [K in ConfigT]: string } = {
+  rider: '_rider.json',
+  bike: '_bike.json',
+  track: '_track.json',
+  powerPlan: '_powerPlan.json',
+};
+
+/**
+ * Split the given configurations object into inidvidual configs and then send them over MQTT
+ *
+ * @param configs dictionary containign the 4 config types (`powerPlan`, `rider`, `track` and `bike`)
+ * @param fileName name of the file that contained all the configs
+ */
+function uploadMultipleConfigs(configs: ConfigBundleT, fileName: string) {
+  // For each config, send the config content over MQTT
+  Object.entries(configs).forEach((configEntry) => {
+    const configType = configEntry[0] as ConfigT;
+    const config = ConfigObjT.check(configEntry[1]);
+
+    // Since this config was uploaded as a bundle give it a different file name to differnetiate it's config type.
+    const file = fileName.replace('.json', configTypeToFileSuffix[configType]);
+
+    sendConfig('upload', configType, file, JSON.stringify(config));
+  });
+
+  toast.success(`Uploaded configs in ${fileName}`);
+}
+
+/**
+ * Read content from the given file and send it on `boost/configs/action` over MQTT.
+ * If the content contains more than one config (i.e. `type` is 'all'), the content is
+ * split into the different configurations before sending.
+ *
+ * @param type the type of the configuration
+ * @param configFile file containing content of the configuration
+ * @param displayErr function to display error if uploaded config is not correct
  */
 export default function uploadConfig(
-  type: BoostConfigType,
+  type: FileConfigT,
   configFile: File,
+  displayErr: (message: string) => void,
 ) {
-  const topic = 'send-config';
   const reader = new FileReader();
 
   // Called when FileReader has completed reading a file
   reader.onload = () => {
-    console.log(reader.result);
-    const payload = {
-      action: 'upload',
-      configType: type,
-      content: reader.result,
-    };
-    emit(topic, JSON.stringify(payload));
+    const fileContent = reader.result as string;
+    const configContent = JSON.parse(fileContent);
+
+    if (!isCorrectContentType(configContent, fileConfigTypeToRuntype[type])) {
+      displayErr(
+        `${configFile.name} is not of the correct type for ${type} config`,
+      );
+    } else if (type === 'bundle') {
+      uploadMultipleConfigs(configContent, configFile.name);
```

Comment:
This function call seems to only exist to reduce the size of the `uploadConfig` function.
If the changes to `sendConfig` were implemented, this function call could be replaced with the remaining body of the function without really increasing the size of this function

Reply:
[Not entertained because I'm unsure of the previous comment]

Reply:
I think this is independent of the `try-catch` discussion as it refers to `sendConfig` and the usefulness of the `uploadMultipleConfigs` function. Again might be a preference thing as I tend to avoid single use functions unless the fact they are functions is being leveraged somehow to make the code better.

I recall from 2102 working with TypeScript and Haskell, that a tutor introduced me to the idea that in strictly typed languages (which TypeScript sort of is), it's actually okay to have large functions/files as the compiler will ensure that changes are propagated properly. Can't find the video I watched on this, but I think it is a thing.

Reply:
Refer to the comment above, `sendConfig` will be used more than once. As for the `uploadMultipleConfigs` I see where you're coming from and I totally agree now (before I used `RunTypes` the function was really too big and the indentations made it harder to read so we decided to split it (you can even find the comment above regrading this). I'm happy to merge them into one function again!

Reply:
I think `sendConfig` should stay a function - there was no question about that. My point was that its existence as a function could be leveraged in the code - reducing the usefulness and necessity of `uploadMultipleConfigs`.

Reply:
So just to make it more clear what part of `uploadMultipleConfigs` could be handled by `sendConfigs` according to you? Happy to chat about this over zoom or discord if messaging isn't that effective.

Reply:
I think talking about it over zoom would be more productive. Maybe @hallgchris can join as well to make sure the things I say actually make sense. Unfortunately, I'm likely not going to be available for the Friday drop-in, but my afternoons are pretty open.